### PR TITLE
Configure email

### PR DESCRIPTION
**Pull request by sslaws**
Configures email queueing and a basic retry batch job.

### CODE REVIEW

**Comment by healthgatewaybc** [comment]
Common Project [SonarQube Report](https://sonarqube-gateway.pathfinder.gov.bc.ca/dashboard?id=commonlibrary_pr_218)

**Comment by healthgatewaybc** [comment]
Patient Project [SonarQube Report](https://sonarqube-gateway.pathfinder.gov.bc.ca/dashboard?id=patientservice_pr_218)

**Comment by healthgatewaybc** [comment]
Medication Project [SonarQube Report](https://sonarqube-gateway.pathfinder.gov.bc.ca/dashboard?id=Medicationservice_pr_218)

**Comment by healthgatewaybc** [comment]
Immunization Project [SonarQube Report](https://sonarqube-gateway.pathfinder.gov.bc.ca/dashboard?id=immunization_pr_218)

**Comment by healthgatewaybc** [comment]
Database Project [SonarQube Report](https://sonarqube-gateway.pathfinder.gov.bc.ca/dashboard?id=databaselibrary_pr_218)

**Comment by healthgatewaybc** [comment]
WebClient Project [SonarQube Report](https://sonarqube-gateway.pathfinder.gov.bc.ca/dashboard?id=webclient_pr_218)

File: Apps/Common/src/Jobs/IEmailJob.cs [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+        /// <summary>
+        /// Sends an email immediately if Priority is standard or higher.
+        /// </summary>
+        /// <param name="emailId">The stored emailId to send.</param>
+        void SendEmail(Guid emailId);
+
+        /// <summary>
+        /// Attempts to send low priority emails.
+        /// </summary>
+        void SendNewLow();
```
Comment (FuriousLlama) [inline Apps/Common/src/Jobs/IEmailJob.cs]: nit: SendNewLow -> SendLowPriorityEmails()

File: Apps/Common/src/Services/EmailQueueService.cs [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+            email.Priority = emailTemplate.Priority;
+            email.Subject = ProcessTemplateString(emailTemplate.Subject, keyValues);
+            email.Body = ProcessTemplateString(emailTemplate.Body, keyValues);
+            email.FormatCode = emailTemplate.FormatCode;
+            return email;
+        }
+
+        private static string ProcessTemplateString(string template, Dictionary<string, string> data)
+        {
+            return Regex.Replace(template, "\\$\\{(.*?)\\}", m =>
```
Comment (FuriousLlama) [inline Apps/Common/src/Services/EmailQueueService.cs]: this looks quite confusing, either lets break it up or add documentation about it is doing

File: Apps/Database/src/Delegates/IEmailDelegate.cs [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+        /// <summary>
+        /// Gets an email in the new state with standard priority or higher.
+        /// </summary>
+        /// <param name="emailId">The Email ID to retrieve.</param>
+        /// <returns>The found Email object.</returns>
+        Email GetNewEmail(Guid emailId);
+
+        /// <summary>
+        /// Gets a list of New low priority emails to process.
```
Comment (FuriousLlama) [inline Apps/Database/src/Delegates/IEmailDelegate.cs]: The implementation also sorts the emails. Having that in the documentation of the interface would be helpful

File: Apps/Database/src/Delegates/IEmailDelegate.cs [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+        /// </summary>
+        /// <param name="maxRows">The maximum amount of emails to return.</param>
+        /// <returns>The list of new low priority emails.</returns>
+        List<Email> GetNewLow(int maxRows);
```
Comment (FuriousLlama) [inline Apps/Database/src/Delegates/IEmailDelegate.cs]: nit: GetNetLow -> GetNewLowPriority

File: Apps/Database/src/Delegates/EmailDelegate.cs [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+                                              p.EmailStatusCode == EmailStatus.New &&
+                                              p.Priority >= EmailPriority.Standard).SingleOrDefault();
+        }
+
+        /// <inheritdoc />
+        public List<Email> GetNewLow(int maxRows)
+        {
+            return this.dbContext.Email.Where(p => p.EmailStatusCode == EmailStatus.New &&
+                                                   p.Priority < EmailPriority.Standard)
+                                        .OrderByDescending(s => s.Priority)
```
Comment (FuriousLlama) [inline Apps/Database/src/Delegates/EmailDelegate.cs]: I dont think this is needed, there is only one priority in the set (low)

File: Apps/Hangfire/src/Jobs/EmailJob.cs [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+            {
+                caught = e;
+                this.logger.LogError($"Unexpected error while sending email {email.Id}", e);
+            }
+#pragma warning restore CA1031 // Restore check
+
+            if (caught != null)
+            {
+                email.LastRetryDateTime = DateTime.UtcNow;
+                email.EmailStatusCode = email.Attempts < 9 ? EmailStatus.Pending : EmailStatus.Error;
```
Comment (FuriousLlama) [inline Apps/Hangfire/src/Jobs/EmailJob.cs]: email.Attempts < 9. Either constant or configuration

File: Apps/Hangfire/src/Jobs/EmailJob.cs [hunk 7]
```diff
@@ -72,10 +74,10 @@ public void SendEmail(Guid emailId)
         }
 
         /// <inheritdoc />
-        public void SendNewLow()
+        public void SendLowPriorityEmails()
         {
             this.logger.LogInformation($"Looking for up to {this.retryFetchSize} low priority emails to send");
-            List<Email> resendEmails = this.emailDelegate.GetNewLow(this.retryFetchSize);
+            List<Email> resendEmails = this.emailDelegate.GeLowPriorityEmail(this.retryFetchSize);
```
Comment (FuriousLlama) [inline Apps/Hangfire/src/Jobs/EmailJob.cs]: typo, GeLowPriorityEmail -> GetLowPriorityEmail

**Pull request merged**